<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Home · HetDister.jl</title><meta name="title" content="Home · HetDister.jl"/><meta property="og:title" content="Home · HetDister.jl"/><meta property="twitter:title" content="Home · HetDister.jl"/><meta name="description" content="Documentation for HetDister.jl."/><meta property="og:description" content="Documentation for HetDister.jl."/><meta property="twitter:description" content="Documentation for HetDister.jl."/><meta property="og:url" content="https://ArndtLab.github.io/HetDister.jl/"/><meta property="twitter:url" content="https://ArndtLab.github.io/HetDister.jl/"/><link rel="canonical" href="https://ArndtLab.github.io/HetDister.jl/"/><script data-outdated-warner src="assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="assets/documenter.js"></script><script src="search_index.js"></script><script src="siteinfo.js"></script><script src="../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href><img src="assets/logo.png" alt="HetDister.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit"><a href>HetDister.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li class="is-active"><a class="tocitem" href>Home</a><ul class="internal"><li><a class="tocitem" href="#Data-form-of-input-and-output"><span>Data form of input and output</span></a></li></ul></li><li><a class="tocitem" href="tutorial/">Tutorial</a></li><li><a class="tocitem" href="diagnostics/">Diagnostics</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Home</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Home</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/ArndtLab/HetDister.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/ArndtLab/HetDister.jl/blob/main/docs/src/index.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="HetDister"><a class="docs-heading-anchor" href="#HetDister">HetDister</a><a id="HetDister-1"></a><a class="docs-heading-anchor-permalink" href="#HetDister" title="Permalink"></a></h1><p>Documentation for <a href="https://github.com/ArndtLab/HetDister.jl">HetDister</a>.</p><p>Module to run demographic inference on diploid genomes, under the assumption of panmixia.</p><h2 id="Data-form-of-input-and-output"><a class="docs-heading-anchor" href="#Data-form-of-input-and-output">Data form of input and output</a><a id="Data-form-of-input-and-output-1"></a><a class="docs-heading-anchor-permalink" href="#Data-form-of-input-and-output" title="Permalink"></a></h2><p>The genome needs to be SNP-called and the genomic distance between consecutive heterozygous positions needs to be computed. Heterozygous positions are the ones with genotype 0/1 or 1/0 (Note that the phase is not important). The input is then a vector containind such distances. Additionally, mutation and recombination rates need to be chosen and passed as input as well. See <a href="tutorial/#Tutorial">Tutorial</a> for more details on preparing input data.</p><p>The demographic model underlying the inference is composed of a variable number of epochs and the population size is constant along each epoch.</p><p>The output is a vector of parameters in the form <code>[L, N0, T1, N1, T2, N2, ...]</code> where <code>L</code> is the total sequence length, <code>N0</code> is the ancestral population size in the furthermost epoch and extending to the infinite past, the subsequent pairs <span>$(T_i, N_i)$</span> are the duration and size of following epochs going from past to present. This format is referred to as <code>TN</code> vector throughout. The length <code>L</code> should match the input sequence length and is floating to improve the fit.</p><ul><li><a href="#HetDister.FitOptions-NTuple{4, Any}"><code>HetDister.FitOptions</code></a></li><li><a href="#HetDister.FitResult"><code>HetDister.FitResult</code></a></li><li><a href="#HetDister.Spectra.laplacekingman-Tuple{Real, Real, AbstractVector{&lt;:Real}}"><code>HetDister.Spectra.laplacekingman</code></a></li><li><a href="#HetDister.Spectra.mldsmcp-NTuple{5, Any}"><code>HetDister.Spectra.mldsmcp</code></a></li><li><a href="#HetDister.adapt_histogram-Tuple{AbstractVector{&lt;:Integer}}"><code>HetDister.adapt_histogram</code></a></li><li><a href="#HetDister.compare_mlds-Tuple{AbstractVector{&lt;:Integer}, AbstractVector{&lt;:Integer}}"><code>HetDister.compare_mlds</code></a></li><li><a href="#HetDister.compare_mlds!-NTuple{4, Any}"><code>HetDister.compare_mlds!</code></a></li><li><a href="#HetDister.compare_models"><code>HetDister.compare_models</code></a></li><li><a href="#HetDister.compute_residuals-Tuple{StatsBase.Histogram, Float64, Float64, Vector}"><code>HetDister.compute_residuals</code></a></li><li><a href="#HetDister.compute_residuals-Tuple{StatsBase.Histogram, StatsBase.Histogram}"><code>HetDister.compute_residuals</code></a></li><li><a href="#HetDister.demoinfer-Tuple{AbstractVector{&lt;:Integer}, AbstractRange{&lt;:Integer}, Float64, Float64}"><code>HetDister.demoinfer</code></a></li><li><a href="#HetDister.demoinfer-Union{Tuple{E}, Tuple{T}, Tuple{StatsBase.Histogram{T, 1, E}, AbstractRange{&lt;:Integer}, FitOptions}} where {T&lt;:Integer, E&lt;:Tuple{AbstractVector{&lt;:Integer}}}"><code>HetDister.demoinfer</code></a></li><li><a href="#HetDister.durations-Tuple{FitResult}"><code>HetDister.durations</code></a></li><li><a href="#HetDister.evd-Tuple{FitResult}"><code>HetDister.evd</code></a></li><li><a href="#HetDister.get_covar-Tuple{FitResult}"><code>HetDister.get_covar</code></a></li><li><a href="#HetDister.get_para-Tuple{FitResult}"><code>HetDister.get_para</code></a></li><li><a href="#HetDister.pop_sizes-Tuple{FitResult}"><code>HetDister.pop_sizes</code></a></li><li><a href="#HetDister.pre_fit!-Union{Tuple{E}, Tuple{T}, Tuple{FitOptions, StatsBase.Histogram{T, 1, E}, Int64}} where {T&lt;:Integer, E&lt;:Tuple{AbstractVector{&lt;:Integer}}}"><code>HetDister.pre_fit!</code></a></li><li><a href="#HetDister.residstructure-Tuple{AbstractVector{&lt;:Real}}"><code>HetDister.residstructure</code></a></li><li><a href="#HetDister.sample_model_epochs!-Union{Tuple{E}, Tuple{T}, Tuple{FitOptions, StatsBase.Histogram{T, 1, E}, AbstractVector{&lt;:Real}}} where {T&lt;:Integer, E&lt;:Tuple{AbstractVector{&lt;:Integer}}}"><code>HetDister.sample_model_epochs!</code></a></li><li><a href="#HetDister.sds-Tuple{FitResult}"><code>HetDister.sds</code></a></li><li><a href="#HetDister.setinit!-Tuple{FitOptions, AbstractVector{Float64}}"><code>HetDister.setinit!</code></a></li></ul><article><details class="docstring" open="true"><summary id="HetDister.FitOptions-NTuple{4, Any}"><a class="docstring-binding" href="#HetDister.FitOptions-NTuple{4, Any}"><code>HetDister.FitOptions</code></a> — <span class="docstring-category">Method</span></summary><section><div><pre><code class="language-julia hljs">FitOptions(Ltot, nhet, mu, rho; kwargs...)</code></pre><p>Construct an an object of type FitOptions, requiring  total genome length <code>Ltot</code> in base pairs, number of heterozygous sites <code>nhet</code>, mutation rate <code>mu</code> and recombination rate <code>rho</code> per base pair per generation.</p><p><strong>Optional Arguments</strong></p><ul><li><code>Tlow::Number=10</code>, <code>Tupp::Number=1e7</code>: The lower and upper bounds for the duration of epochs.</li><li><code>Nlow::Number=10</code>, <code>Nupp::Number=1e8</code>: The lower and upper bounds for the population sizes.</li><li><code>level::Float64=0.95</code>: The confidence level for the confidence intervals on the parameters estimates.</li><li><code>solver</code>: The solver to use for the optimization, default is <code>LBFGS()</code>.</li><li><code>force::Bool=true</code>: if true try to fit further epochs even when no signal is found.</li><li><code>maxnts::Int=5</code>: The maximum number of new time splits to consider when adding a new epoch. Higher is greedier.</li><li><code>order::Int=0</code>: maximum number of higher order SMC&#39; corrections to account for (i.e. number of intermediate recombination events plus one). When zero, it is set automatically.</li><li><code>ndt::Int=0</code>: number of Legendre nodes to use for numerical integration. When zero, it is set automatically.</li><li><code>locut::Int=1</code>: index of the first histogram bin to consider in the fit.</li></ul><p><strong>Optim Arguments</strong></p><p>Additional keywords are passed to <code>Optimization.solve</code>, see <a href="https://docs.sciml.ai/Optimization/stable/API/solve/#Common-Solver-Options-(Solve-Keyword-Arguments)">Optimization.jl</a>. and the specific <code>Optim.jl</code> section, which is the default optimizer. Defaults are:</p><ul><li><code>maxiters = 6000</code></li><li><code>maxtime = 60</code></li><li><code>g_tol = 5e-8</code></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ArndtLab/HetDister.jl/blob/e4e412c46eaf12b23898de36c010646a1a57ee9e/src/utils.jl#L195-L225">source</a></section></details></article><article><details class="docstring" open="true"><summary id="HetDister.FitResult"><a class="docstring-binding" href="#HetDister.FitResult"><code>HetDister.FitResult</code></a> — <span class="docstring-category">Type</span></summary><section><div><pre><code class="language-julia hljs">struct FitResult</code></pre><p>A data structure to store the results of a fit.</p><p>See the introduction for how the model is  parameterized <a href="#Data-form-of-input-and-output">Data form of input and output</a>. Some methods are defined for this type to get the vector of parameters, std errors,  model evidence, etc. See <a href="#HetDister.get_para-Tuple{FitResult}"><code>get_para</code></a>, <a href="#HetDister.sds-Tuple{FitResult}"><code>sds</code></a>, <a href="#HetDister.evd-Tuple{FitResult}"><code>evd</code></a>,  <a href="#HetDister.pop_sizes-Tuple{FitResult}"><code>pop_sizes</code></a>, <a href="#HetDister.durations-Tuple{FitResult}"><code>durations</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ArndtLab/HetDister.jl/blob/e4e412c46eaf12b23898de36c010646a1a57ee9e/src/utils.jl#L1-L11">source</a></section></details></article><article><details class="docstring" open="true"><summary id="HetDister.adapt_histogram-Tuple{AbstractVector{&lt;:Integer}}"><a class="docstring-binding" href="#HetDister.adapt_histogram-Tuple{AbstractVector{&lt;:Integer}}"><code>HetDister.adapt_histogram</code></a> — <span class="docstring-category">Method</span></summary><section><div><pre><code class="language-julia hljs">adapt_histogram(segments::AbstractVector{&lt;:Integer}; lo::Int=1, hi::Int=50_000_000, nbins::Int=0, tailthr::Int=0)</code></pre><p>Build an histogram from <code>segments</code> logbinned between <code>lo</code> and <code>hi</code> with <code>nbins</code> bins. <code>nbins</code> is automatically determined by default.</p><p>The upper limit is adapted to ensure logspacing with the requested <code>nbins</code>. The adaptive strategy is such that the last bin has at least <code>tailthr</code> segments.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ArndtLab/HetDister.jl/blob/e4e412c46eaf12b23898de36c010646a1a57ee9e/src/HetDister.jl#L128-L136">source</a></section></details></article><article><details class="docstring" open="true"><summary id="HetDister.compare_mlds!-NTuple{4, Any}"><a class="docstring-binding" href="#HetDister.compare_mlds!-NTuple{4, Any}"><code>HetDister.compare_mlds!</code></a> — <span class="docstring-category">Method</span></summary><section><div><pre><code class="language-julia hljs">compare_mlds!(h1::Histogram, h2::Histogram, theta1::Float64, theta2::Float64)</code></pre><p>The same as <code>compare_mlds</code>, except that it takes two histograms and mutates them. Return values are the same as <code>compare_mlds</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ArndtLab/HetDister.jl/blob/e4e412c46eaf12b23898de36c010646a1a57ee9e/src/HetDister.jl#L185-L190">source</a></section></details></article><article><details class="docstring" open="true"><summary id="HetDister.compare_mlds-Tuple{AbstractVector{&lt;:Integer}, AbstractVector{&lt;:Integer}}"><a class="docstring-binding" href="#HetDister.compare_mlds-Tuple{AbstractVector{&lt;:Integer}, AbstractVector{&lt;:Integer}}"><code>HetDister.compare_mlds</code></a> — <span class="docstring-category">Method</span></summary><section><div><pre><code class="language-julia hljs">compare_mlds(segs1::Vector{Int}, segs2::Vector{Int}; lo = 1, hi = 1_000_000, nbins = 200)</code></pre><p>Build two histograms from <code>segs1</code> and <code>segs2</code>, rescale them for number and average heterozygosity and return four vectors containing respectively common midpoints for bins, the two rescaled weights and variances of the difference between weights.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ArndtLab/HetDister.jl/blob/e4e412c46eaf12b23898de36c010646a1a57ee9e/src/HetDister.jl#L167-L174">source</a></section></details></article><article><details class="docstring" open="true"><summary id="HetDister.compare_models"><a class="docstring-binding" href="#HetDister.compare_models"><code>HetDister.compare_models</code></a> — <span class="docstring-category">Function</span></summary><section><div><pre><code class="language-julia hljs">compare_models(models[, mask])</code></pre><p>Compare the models parameterized by <code>FitResult</code>s and return the best one. Takes an iterable of <code>FitResult</code> as input and optionally a boolean mask to reflect prior knowledge on models to discard.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ArndtLab/HetDister.jl/blob/e4e412c46eaf12b23898de36c010646a1a57ee9e/src/corrections.jl#L179-L185">source</a></section></details></article><article><details class="docstring" open="true"><summary id="HetDister.compute_residuals-Tuple{StatsBase.Histogram, Float64, Float64, Vector}"><a class="docstring-binding" href="#HetDister.compute_residuals-Tuple{StatsBase.Histogram, Float64, Float64, Vector}"><code>HetDister.compute_residuals</code></a> — <span class="docstring-category">Method</span></summary><section><div><pre><code class="language-julia hljs">compute_residuals(h::Histogram, mu, rho, TN::Vector; naive=false)</code></pre><p>Compute the residuals between the observed and expected weights.</p><p><strong>Optional arguments</strong></p><ul><li><code>naive::Bool=false</code>: if true the expected weights are computed using the closed form integral, otherwise using higher order transition probabilities from SMC&#39; theory.</li><li><code>order::Int=10</code>: maximum number of higher order corrections to use when <code>naive</code> is false, i.e. number of intermediate recombination events plus one.</li><li><code>ndt::Int=800</code>: number of Legendre nodes to use when <code>naive</code> is false.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ArndtLab/HetDister.jl/blob/e4e412c46eaf12b23898de36c010646a1a57ee9e/src/HetDister.jl#L47-L59">source</a></section></details></article><article><details class="docstring" open="true"><summary id="HetDister.compute_residuals-Tuple{StatsBase.Histogram, StatsBase.Histogram}"><a class="docstring-binding" href="#HetDister.compute_residuals-Tuple{StatsBase.Histogram, StatsBase.Histogram}"><code>HetDister.compute_residuals</code></a> — <span class="docstring-category">Method</span></summary><section><div><pre><code class="language-julia hljs">compute_residuals(h1::Histogram, h2::Histogram)</code></pre><p>When two histograms are given the std error of the difference is used for normalization.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ArndtLab/HetDister.jl/blob/e4e412c46eaf12b23898de36c010646a1a57ee9e/src/HetDister.jl#L76-L81">source</a></section></details></article><article><details class="docstring" open="true"><summary id="HetDister.demoinfer-Tuple{AbstractVector{&lt;:Integer}, AbstractRange{&lt;:Integer}, Float64, Float64}"><a class="docstring-binding" href="#HetDister.demoinfer-Tuple{AbstractVector{&lt;:Integer}, AbstractRange{&lt;:Integer}, Float64, Float64}"><code>HetDister.demoinfer</code></a> — <span class="docstring-category">Method</span></summary><section><div><pre><code class="language-julia hljs">demoinfer(segments::AbstractVector{&lt;:Integer}, epochrange::AbstractRange{&lt;:Integer}, mu::Float64, rho::Float64; kwargs...)</code></pre><p>Make an histogram with IBS <code>segments</code> and infer demographic models with piece-wise constant epochs where the number of epochs is <code>epochrange</code>.</p><p>Return a named tuple which contains the fields:</p><ul><li><code>fits</code>: a vector of <code>FitResult</code> (see <a href="#HetDister.FitResult"><code>FitResult</code></a>)</li><li><code>chains</code>: a vector of vectors of <code>FitResult</code>, one for each iteration of the correction procedure, and one chain per model</li><li><code>corrections</code>: a vector of vectors of corrections, one for each iteration of the correction procedure, and one vector of corrections per model. Corrections are histogram counts, therefore they have the same shape.</li><li><code>h_obs</code>: the histogram of the observed segments</li><li><code>h_mods</code>: a vector of modified histograms, one for each model, with higher order corrections applied.</li><li><code>yth</code>: a vector of vectors of the expected weights, one for each model</li><li><code>deltas</code>: a vector of vectors of the maximum absolute difference between corrections in consecutive iterations, and for each model.</li><li><code>conv</code>: a vector of booleans, one for each model, indicating whether the maximum iterations were reached (false) or whether the procedure  converged before (true).</li></ul><p><strong>Optional Arguments</strong></p><ul><li><code>fop::FitOptions = FitOptions(sum(segments), mu, rho)</code>: the fit options, see <a href="#HetDister.FitOptions-NTuple{4, Any}"><code>FitOptions</code></a>.</li><li><code>lo::Int=1</code>: The lowest segment length to be considered in the histogram</li><li><code>hi::Int=50_000_000</code>: The highest segment length to be considered in the histogram</li><li><code>nbins::Int=fop.ndt</code>: The number of bins to use in the histogram</li><li><code>iters::Int=20</code>: The number of iterations to perform. It might converge earlier</li><li><code>reltol::Float64=1e-2</code>: The relative tolerance to use for convergence, i.e. the maximum absolute difference between corrections in consecutive iterations.</li><li><code>corcut::Int=fop.locut-1</code>: The index of the last histogram bin to apply corrections to.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ArndtLab/HetDister.jl/blob/e4e412c46eaf12b23898de36c010646a1a57ee9e/src/corrections.jl#L6-L39">source</a></section></details></article><article><details class="docstring" open="true"><summary id="HetDister.demoinfer-Union{Tuple{E}, Tuple{T}, Tuple{StatsBase.Histogram{T, 1, E}, AbstractRange{&lt;:Integer}, FitOptions}} where {T&lt;:Integer, E&lt;:Tuple{AbstractVector{&lt;:Integer}}}"><a class="docstring-binding" href="#HetDister.demoinfer-Union{Tuple{E}, Tuple{T}, Tuple{StatsBase.Histogram{T, 1, E}, AbstractRange{&lt;:Integer}, FitOptions}} where {T&lt;:Integer, E&lt;:Tuple{AbstractVector{&lt;:Integer}}}"><code>HetDister.demoinfer</code></a> — <span class="docstring-category">Method</span></summary><section><div><pre><code class="language-julia hljs">demoinfer(h::Histogram, epochrange, fop::FitOptions; iters=20, reltol=1e-2, corcut=fop.locut-1, finalize=false)
demoinfer(h, epochs, fop; iters=20, reltol=1e-2, corcut=fop.locut-1, finalize=false)</code></pre><p>Take an histogram of IBS segments, fit options, and infer demographic models with piece-wise constant epochs where the number of epochs is <code>epochrange</code>. Return a named tuple as above.</p><p>If <code>epochrange</code> is a integer, then it fits only the model with that number of epochs. In this case the returned named tuple contains only one element per field, instead of a vector.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ArndtLab/HetDister.jl/blob/e4e412c46eaf12b23898de36c010646a1a57ee9e/src/corrections.jl#L53-L63">source</a></section></details></article><article><details class="docstring" open="true"><summary id="HetDister.durations-Tuple{FitResult}"><a class="docstring-binding" href="#HetDister.durations-Tuple{FitResult}"><code>HetDister.durations</code></a> — <span class="docstring-category">Method</span></summary><section><div><pre><code class="language-julia hljs">durations(fit::FitResult)</code></pre><p>Return the fitted durations of the epochs.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ArndtLab/HetDister.jl/blob/e4e412c46eaf12b23898de36c010646a1a57ee9e/src/utils.jl#L67-L71">source</a></section></details></article><article><details class="docstring" open="true"><summary id="HetDister.evd-Tuple{FitResult}"><a class="docstring-binding" href="#HetDister.evd-Tuple{FitResult}"><code>HetDister.evd</code></a> — <span class="docstring-category">Method</span></summary><section><div><pre><code class="language-julia hljs">evd(fit::FitResult)</code></pre><p>Return the evidence of the fit.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ArndtLab/HetDister.jl/blob/e4e412c46eaf12b23898de36c010646a1a57ee9e/src/utils.jl#L53-L57">source</a></section></details></article><article><details class="docstring" open="true"><summary id="HetDister.get_covar-Tuple{FitResult}"><a class="docstring-binding" href="#HetDister.get_covar-Tuple{FitResult}"><code>HetDister.get_covar</code></a> — <span class="docstring-category">Method</span></summary><section><div><pre><code class="language-julia hljs">get_covar(fit::FitResult)</code></pre><p>Return the covariance matrix of the parameters of the fit, computed as the inverse of the likelihood Hessian at the optimum.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ArndtLab/HetDister.jl/blob/e4e412c46eaf12b23898de36c010646a1a57ee9e/src/utils.jl#L76-L80">source</a></section></details></article><article><details class="docstring" open="true"><summary id="HetDister.get_para-Tuple{FitResult}"><a class="docstring-binding" href="#HetDister.get_para-Tuple{FitResult}"><code>HetDister.get_para</code></a> — <span class="docstring-category">Method</span></summary><section><div><pre><code class="language-julia hljs">pars(fit::FitResult)</code></pre><p>Return the parameters of the fit.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ArndtLab/HetDister.jl/blob/e4e412c46eaf12b23898de36c010646a1a57ee9e/src/utils.jl#L39-L43">source</a></section></details></article><article><details class="docstring" open="true"><summary id="HetDister.pop_sizes-Tuple{FitResult}"><a class="docstring-binding" href="#HetDister.pop_sizes-Tuple{FitResult}"><code>HetDister.pop_sizes</code></a> — <span class="docstring-category">Method</span></summary><section><div><pre><code class="language-julia hljs">pop_sizes(fit::FitResult)</code></pre><p>Return the fitted population sizes, from past to present.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ArndtLab/HetDister.jl/blob/e4e412c46eaf12b23898de36c010646a1a57ee9e/src/utils.jl#L60-L64">source</a></section></details></article><article><details class="docstring" open="true"><summary id="HetDister.pre_fit!-Union{Tuple{E}, Tuple{T}, Tuple{FitOptions, StatsBase.Histogram{T, 1, E}, Int64}} where {T&lt;:Integer, E&lt;:Tuple{AbstractVector{&lt;:Integer}}}"><a class="docstring-binding" href="#HetDister.pre_fit!-Union{Tuple{E}, Tuple{T}, Tuple{FitOptions, StatsBase.Histogram{T, 1, E}, Int64}} where {T&lt;:Integer, E&lt;:Tuple{AbstractVector{&lt;:Integer}}}"><code>HetDister.pre_fit!</code></a> — <span class="docstring-category">Method</span></summary><section><div><pre><code class="language-julia hljs">pre_fit!(fop::FitOptions, h::Histogram, nfits)</code></pre><p>Preliminarily fit <code>h</code> with an approximate model of piece-wise constant  epochs for each number of epochs from 1 to <code>nfits</code>.</p><p>See <a href="#HetDister.FitOptions-NTuple{4, Any}"><code>FitOptions</code></a> for how to specify them. It modifies <code>fop</code> in place to adapt it to all the requested epochs. Return a vector of <code>FitResult</code>, one for each number of epochs, see also <a href="#HetDister.FitResult"><code>FitResult</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ArndtLab/HetDister.jl/blob/e4e412c46eaf12b23898de36c010646a1a57ee9e/src/sequential_fit.jl#L90-L101">source</a></section></details></article><article><details class="docstring" open="true"><summary id="HetDister.residstructure-Tuple{AbstractVector{&lt;:Real}}"><a class="docstring-binding" href="#HetDister.residstructure-Tuple{AbstractVector{&lt;:Real}}"><code>HetDister.residstructure</code></a> — <span class="docstring-category">Method</span></summary><section><div><pre><code class="language-julia hljs">residstructure(residuals::AbstractVector{&lt;:Real}; frame::Int = length(residuals)÷20)</code></pre><p>Compute the p-values for the correlation between adjacent residuals in a sliding window of size <code>frame</code>. The p-value is the right tail of the t-distribution.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ArndtLab/HetDister.jl/blob/e4e412c46eaf12b23898de36c010646a1a57ee9e/src/HetDister.jl#L92-L97">source</a></section></details></article><article><details class="docstring" open="true"><summary id="HetDister.sample_model_epochs!-Union{Tuple{E}, Tuple{T}, Tuple{FitOptions, StatsBase.Histogram{T, 1, E}, AbstractVector{&lt;:Real}}} where {T&lt;:Integer, E&lt;:Tuple{AbstractVector{&lt;:Integer}}}"><a class="docstring-binding" href="#HetDister.sample_model_epochs!-Union{Tuple{E}, Tuple{T}, Tuple{FitOptions, StatsBase.Histogram{T, 1, E}, AbstractVector{&lt;:Real}}} where {T&lt;:Integer, E&lt;:Tuple{AbstractVector{&lt;:Integer}}}"><code>HetDister.sample_model_epochs!</code></a> — <span class="docstring-category">Method</span></summary><section><div><pre><code class="language-julia hljs">sample_model_epochs!(options::FitOptions, h::Histogram{T,1,E}, init::AbstractVector{&lt;:Real}; nsamples = 10_000, findmode = false)</code></pre><p>Sample <code>nsamples</code> from the posterior distribution of the parameters, starting from initial point <code>init</code>.</p><p>Requires the observed histogram <code>h</code> and the fit options <code>options</code>. Return a <code>Chains</code> object from the <code>MCMCDiagnostics</code> module of <code>Turing</code>, which contains the samples from the posterior distribution. If <code>findmode</code> is true, the function will first find the mode of the posterior distribution using optimization, and then use that as the initial point for sampling. Otherwise, it will use the provided <code>init</code> as the initial point for sampling.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ArndtLab/HetDister.jl/blob/e4e412c46eaf12b23898de36c010646a1a57ee9e/src/mle_optimization.jl#L199-L212">source</a></section></details></article><article><details class="docstring" open="true"><summary id="HetDister.sds-Tuple{FitResult}"><a class="docstring-binding" href="#HetDister.sds-Tuple{FitResult}"><code>HetDister.sds</code></a> — <span class="docstring-category">Method</span></summary><section><div><pre><code class="language-julia hljs">sds(fit::FitResult)</code></pre><p>Return the standard deviations of the parameters of the fit.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ArndtLab/HetDister.jl/blob/e4e412c46eaf12b23898de36c010646a1a57ee9e/src/utils.jl#L46-L50">source</a></section></details></article><article><details class="docstring" open="true"><summary id="HetDister.setinit!-Tuple{FitOptions, AbstractVector{Float64}}"><a class="docstring-binding" href="#HetDister.setinit!-Tuple{FitOptions, AbstractVector{Float64}}"><code>HetDister.setinit!</code></a> — <span class="docstring-category">Method</span></summary><section><div><pre><code class="language-julia hljs">setinit!(fop::FitOptions, init::AbstractVector{Float64})</code></pre><p>Set the initial vector of parameters for the optimization which takes the <code>FitOptions</code> object <code>fop</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ArndtLab/HetDister.jl/blob/e4e412c46eaf12b23898de36c010646a1a57ee9e/src/utils.jl#L302-L306">source</a></section></details></article><article><details class="docstring" open="true"><summary id="HetDister.Spectra.laplacekingman-Tuple{Real, Real, AbstractVector{&lt;:Real}}"><a class="docstring-binding" href="#HetDister.Spectra.laplacekingman-Tuple{Real, Real, AbstractVector{&lt;:Real}}"><code>HetDister.Spectra.laplacekingman</code></a> — <span class="docstring-category">Method</span></summary><section><div><pre><code class="language-julia hljs">laplacekingman(r, mu, TN)</code></pre><p>Compute the approximate number of segments of length <code>r</code>  using the Laplace transform of the Kingman coalescent at frequency <code>2mu r</code>, given mutation rate <code>mu</code> and population size history <code>TN</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ArndtLab/HetDister.jl/blob/e4e412c46eaf12b23898de36c010646a1a57ee9e/src/Spectra/Spectra.jl#L63-L69">source</a></section></details></article><article><details class="docstring" open="true"><summary id="HetDister.Spectra.mldsmcp-NTuple{5, Any}"><a class="docstring-binding" href="#HetDister.Spectra.mldsmcp-NTuple{5, Any}"><code>HetDister.Spectra.mldsmcp</code></a> — <span class="docstring-category">Method</span></summary><section><div><pre><code class="language-julia hljs">mldsmcp(rs, edges, mu, rho, TN; order = 10, ndt = 800)</code></pre><p>Compute the expected number of segments with length in each bin defined by <code>edges</code>, given the midpoints <code>rs</code>, mutation rate <code>mu</code>, recombination rate <code>rho</code>, and population size history <code>TN</code>.</p><p>The computation uses the SMC&#39; higher order transition probabilities with <code>order</code> maximum number of intermediate recombination events plus one, and <code>ndt</code> Legendre nodes for the numerical integration.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ArndtLab/HetDister.jl/blob/e4e412c46eaf12b23898de36c010646a1a57ee9e/src/Spectra/Spectra.jl#L20-L30">source</a></section></details></article></article><nav class="docs-footer"><a class="docs-footer-nextpage" href="tutorial/">Tutorial »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.17.0 on <span class="colophon-date" title="Monday 23 February 2026 16:59">Monday 23 February 2026</span>. Using Julia version 1.12.5.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
