<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Home · DemoInfer.jl</title><meta name="title" content="Home · DemoInfer.jl"/><meta property="og:title" content="Home · DemoInfer.jl"/><meta property="twitter:title" content="Home · DemoInfer.jl"/><meta name="description" content="Documentation for DemoInfer.jl."/><meta property="og:description" content="Documentation for DemoInfer.jl."/><meta property="twitter:description" content="Documentation for DemoInfer.jl."/><meta property="og:url" content="https://ArndtLab.github.io/DemoInfer.jl/"/><meta property="twitter:url" content="https://ArndtLab.github.io/DemoInfer.jl/"/><link rel="canonical" href="https://ArndtLab.github.io/DemoInfer.jl/"/><script data-outdated-warner src="assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="assets/documenter.js"></script><script src="search_index.js"></script><script src="siteinfo.js"></script><script src="../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href><img src="assets/logo.png" alt="DemoInfer.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit"><a href>DemoInfer.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li class="is-active"><a class="tocitem" href>Home</a><ul class="internal"><li><a class="tocitem" href="#Data-form-of-input-and-output"><span>Data form of input and output</span></a></li></ul></li><li><a class="tocitem" href="tutorial/">Tutorial</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Home</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Home</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/ArndtLab/DemoInfer.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/ArndtLab/DemoInfer.jl/blob/main/docs/src/index.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="DemoInfer"><a class="docs-heading-anchor" href="#DemoInfer">DemoInfer</a><a id="DemoInfer-1"></a><a class="docs-heading-anchor-permalink" href="#DemoInfer" title="Permalink"></a></h1><p>Documentation for <a href="https://github.com/ArndtLab/DemoInfer.jl">DemoInfer</a>.</p><p>Module to run demographic inference on diploid genomes, under the assumption of panmixia (i.e. the inferred effective population size is half the inverse of the observed mean coalescence rate). See <a href="https://github.com/ArndtLab/DemoInferDemo">this</a> repo for a demo of how to use it.</p><h2 id="Data-form-of-input-and-output"><a class="docs-heading-anchor" href="#Data-form-of-input-and-output">Data form of input and output</a><a id="Data-form-of-input-and-output-1"></a><a class="docs-heading-anchor-permalink" href="#Data-form-of-input-and-output" title="Permalink"></a></h2><p>The genome needs to be SNP-called and the genomic distance between consecutive heterozygous positions needs to be computed. Heterozygous positions are the ones with genotype 0/1 or 1/0 (Note that the phase is not important). The input is then a vector containind such distances. Additionally, mutation and recombination rates need to be chosen and passed as input as well.</p><p>For example, suppose you have a <code>.vcf</code> file with called variants you want to analyze. Then you may compute distances between heterozygous SNPs as follows:</p><pre><code class="language-julia hljs">using CSV
using DataFrames
using DataFramesMeta

f = &quot;/myproject/myfavouritespecies.vcf&quot;
df = CSV.read(f, DataFrame, 
    delim=&#39;\t&#39;, 
    comment=&quot;##&quot;,
    missingstring=[&quot;.&quot;, &quot;NaN&quot;],
    normalizenames=true,
    ntasks = 1,
    drop = [:INFO, :ID, :FILTER],
)

# remove homozygous variants
@chain df begin
    @rsubset! (:SampleName[1] == &#39;1&#39; &amp;&amp; :SampleName[3] == &#39;0&#39;) || (:SampleName[1] == &#39;0&#39; &amp;&amp; :SampleName[3] == &#39;1&#39;)
end

ils = df.POS[2:end] .- df.POS[1:end-1]
@assert all(ils .&gt; 0)</code></pre><p>The demographic model underlying the inference is composed of a variable number of epochs and the population size is constant along each epoch.</p><p>The output is a vector of parameters in the form <code>[L, N0, T1, N1, T2, N2, ...]</code> where <code>L</code> is the genome length, <code>N0</code> is the ancestral population size in the furthermost epoch and extending to the infinite past, the subsequent pairs <span>$(T_i, N_i)$</span> are the duration and size of following epochs going from past to present. This format is referred to as <code>TN</code> vector throughout.</p><ul><li><a href="#DemoInfer.FitOptions-Tuple{Any, Any, Any}"><code>DemoInfer.FitOptions</code></a></li><li><a href="#DemoInfer.FitResult"><code>DemoInfer.FitResult</code></a></li><li><a href="#DemoInfer.adapt_histogram-Tuple{AbstractVector{&lt;:Integer}}"><code>DemoInfer.adapt_histogram</code></a></li><li><a href="#DemoInfer.compare_mlds-Tuple{AbstractVector{&lt;:Integer}, AbstractVector{&lt;:Integer}}"><code>DemoInfer.compare_mlds</code></a></li><li><a href="#DemoInfer.compare_models-Tuple{Any}"><code>DemoInfer.compare_models</code></a></li><li><a href="#DemoInfer.compute_residuals-Tuple{StatsBase.Histogram, Float64, Float64, Vector}"><code>DemoInfer.compute_residuals</code></a></li><li><a href="#DemoInfer.compute_residuals-Tuple{StatsBase.Histogram, StatsBase.Histogram}"><code>DemoInfer.compute_residuals</code></a></li><li><a href="#DemoInfer.demoinfer-Union{Tuple{E}, Tuple{T}, Tuple{StatsBase.Histogram{T, 1, E}, Int64, FitOptions}} where {T&lt;:Integer, E&lt;:Tuple{AbstractVector{&lt;:Integer}}}"><code>DemoInfer.demoinfer</code></a></li><li><a href="#DemoInfer.demoinfer-Tuple{AbstractVector{&lt;:Integer}, Int64, Float64, Float64}"><code>DemoInfer.demoinfer</code></a></li><li><a href="#DemoInfer.durations-Tuple{FitResult}"><code>DemoInfer.durations</code></a></li><li><a href="#DemoInfer.evd-Tuple{FitResult}"><code>DemoInfer.evd</code></a></li><li><a href="#DemoInfer.get_para-Tuple{FitResult}"><code>DemoInfer.get_para</code></a></li><li><a href="#DemoInfer.pop_sizes-Tuple{FitResult}"><code>DemoInfer.pop_sizes</code></a></li><li><a href="#DemoInfer.pre_fit-Union{Tuple{E}, Tuple{T}, Tuple{StatsBase.Histogram{T, 1, E}, Int64, Float64, Float64, Int64, Int64, Number}} where {T&lt;:Integer, E&lt;:Tuple{AbstractVector{&lt;:Integer}}}"><code>DemoInfer.pre_fit</code></a></li><li><a href="#DemoInfer.sds-Tuple{FitResult}"><code>DemoInfer.sds</code></a></li></ul><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="DemoInfer.FitOptions-Tuple{Any, Any, Any}" href="#DemoInfer.FitOptions-Tuple{Any, Any, Any}"><code>DemoInfer.FitOptions</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">FitOptions(Ltot, mu, rho; kwargs...)</code></pre><p>Construct an an object of type FitOptions, requiring  total genome length <code>Ltot</code> in base pairs, mutation rate and recombination rate per base pair per generation.</p><p><strong>Optional Arguments</strong></p><ul><li><code>Tlow::Number=10</code>, <code>Tupp::Number=1e7</code>: The lower and upper bounds for the duration of epochs.</li><li><code>Nlow::Number=10</code>, <code>Nupp::Number=1e8</code>: The lower and upper bounds for the population sizes.</li><li><code>level::Float64=0.95</code>: The confidence level for the confidence intervals on the parameters estimates.</li><li><code>solver</code>: The solver to use for the optimization, default is <code>LBFGS()</code>.</li><li><code>opt</code>: The optimization options, a named tuple which is passed to </li></ul><p><a href="https://github.com/JuliaNLSolvers/Optim.jl">Optim.jl</a>. Default has keywords:     - <code>iterations = 6000</code>     - <code>allow_f_increases = true</code>     - <code>time_limit = 60</code>     - <code>g_tol = 5e-8</code>     - <code>show_warnings = false</code>.</p><ul><li><code>smallest_segment::Int=1</code>: The smallest segment size present in the histogram to consider </li></ul><p>for the signal search.</p><ul><li><code>force::Bool=true</code>: if true try to fit further epochs even when no signal is found.</li><li><code>maxnts::Int=10</code>: The maximum number of new time splits to consider when adding a new epoch.</li></ul><p>Higher is greedier.</p><ul><li><code>naive::Bool=true</code>: if true the expected weights are computed using the closed form integral, otherwise using higher order transition probabilities from SMC&#39; theory (slower).</li><li><code>order::Int=10</code>: maximum number of higher order corrections to use when <code>naive</code> is false, i.e. number of intermediate recombination events plus one.</li><li><code>ndt::Int=800</code>: number of Legendre nodes to use when <code>naive</code> is false.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ArndtLab/DemoInfer.jl/blob/85763d2f4718581c9e647262223d36ddd81445e6/src/utils.jl#L175-L206">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="DemoInfer.FitResult" href="#DemoInfer.FitResult"><code>DemoInfer.FitResult</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">struct FitResult</code></pre><p>A data structure to store the results of a fit.</p><p>See the introduction for how the model is  parameterized <a href="#Data-form-of-input-and-output">Data form of input and output</a>. Some methods are defined for this type to get the vector of parameters, std errors,  model evidence, etc. See <a href="#DemoInfer.get_para-Tuple{FitResult}"><code>get_para</code></a>, <a href="#DemoInfer.sds-Tuple{FitResult}"><code>sds</code></a>, <a href="#DemoInfer.evd-Tuple{FitResult}"><code>evd</code></a>,  <a href="#DemoInfer.pop_sizes-Tuple{FitResult}"><code>pop_sizes</code></a>, <a href="#DemoInfer.durations-Tuple{FitResult}"><code>durations</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ArndtLab/DemoInfer.jl/blob/85763d2f4718581c9e647262223d36ddd81445e6/src/utils.jl#L1-L11">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="DemoInfer.adapt_histogram-Tuple{AbstractVector{&lt;:Integer}}" href="#DemoInfer.adapt_histogram-Tuple{AbstractVector{&lt;:Integer}}"><code>DemoInfer.adapt_histogram</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">adapt_histogram(segments::AbstractVector{&lt;:Integer}; lo::Int=1, hi::Int=50_000_000, nbins::Int=200)</code></pre><p>Build an histogram from <code>segments</code> logbinned between <code>lo</code> and <code>hi</code> with <code>nbins</code> bins (see <code>HistogramBinnings.jl</code>).</p><p>The upper limit is adapted to the maximum observed length, so default value is on purpose high.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ArndtLab/DemoInfer.jl/blob/85763d2f4718581c9e647262223d36ddd81445e6/src/DemoInfer.jl#L104-L112">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="DemoInfer.compare_mlds-Tuple{AbstractVector{&lt;:Integer}, AbstractVector{&lt;:Integer}}" href="#DemoInfer.compare_mlds-Tuple{AbstractVector{&lt;:Integer}, AbstractVector{&lt;:Integer}}"><code>DemoInfer.compare_mlds</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">compare_mlds(segs1::Vector{Int}, segs2::Vector{Int}; lo = 1, hi = 1_000_000, nbins = 200)</code></pre><p>Build two histograms from <code>segs1</code> and <code>segs2</code>, rescale them for number and average heterozygosity and return four vectors containing respectively common midpoints for bins, the two rescaled weights and variances of the difference between weights.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ArndtLab/DemoInfer.jl/blob/85763d2f4718581c9e647262223d36ddd81445e6/src/DemoInfer.jl#L152-L159">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="DemoInfer.compare_models-Tuple{Any}" href="#DemoInfer.compare_models-Tuple{Any}"><code>DemoInfer.compare_models</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">compare_models(models)</code></pre><p>Compare the models parameterized by <code>FitResult</code>s and return the best one. Takes an iterable of <code>FitResult</code> as input.</p><p><strong>Theoretical explanation</strong></p><p>TBD</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ArndtLab/DemoInfer.jl/blob/85763d2f4718581c9e647262223d36ddd81445e6/src/corrections.jl#L107-L115">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="DemoInfer.compute_residuals-Tuple{StatsBase.Histogram, Float64, Float64, Vector}" href="#DemoInfer.compute_residuals-Tuple{StatsBase.Histogram, Float64, Float64, Vector}"><code>DemoInfer.compute_residuals</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">compute_residuals(h::Histogram, mu, rho, TN::Vector; naive=false)</code></pre><p>Compute the residuals between the observed and expected weights.</p><p><strong>Optional arguments</strong></p><ul><li><code>naive::Bool=false</code>: if true the expected weights are computed using the closed form integral, otherwise using higher order transition probabilities from SMC&#39; theory.</li><li><code>order::Int=10</code>: maximum number of higher order corrections to use when <code>naive</code> is false, i.e. number of intermediate recombination events plus one.</li><li><code>ndt::Int=800</code>: number of Legendre nodes to use when <code>naive</code> is false.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ArndtLab/DemoInfer.jl/blob/85763d2f4718581c9e647262223d36ddd81445e6/src/DemoInfer.jl#L42-L54">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="DemoInfer.compute_residuals-Tuple{StatsBase.Histogram, StatsBase.Histogram}" href="#DemoInfer.compute_residuals-Tuple{StatsBase.Histogram, StatsBase.Histogram}"><code>DemoInfer.compute_residuals</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">compute_residuals(h1::Histogram, h2::Histogram)</code></pre><p>When two histograms are given the std error of the difference is used for normalization.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ArndtLab/DemoInfer.jl/blob/85763d2f4718581c9e647262223d36ddd81445e6/src/DemoInfer.jl#L71-L76">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="DemoInfer.demoinfer-Tuple{AbstractVector{&lt;:Integer}, Int64, Float64, Float64}" href="#DemoInfer.demoinfer-Tuple{AbstractVector{&lt;:Integer}, Int64, Float64, Float64}"><code>DemoInfer.demoinfer</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">demoinfer(segments::AbstractVector{&lt;:Integer}, epochs::Int, mu::Float64, rho::Float64; kwargs...)</code></pre><p>Make an histogram with IBS <code>segments</code> and infer demographic histories with piece-wise constant epochs where the number of epochs is smaller or equal to <code>epochs</code>.</p><p>Return a vector of <code>FitResult</code> of length smaller or equal to <code>epochs</code>,  see <a href="#DemoInfer.FitResult"><code>FitResult</code></a>, <code>mu</code> and <code>rho</code> are respectively the mutation  and recombination rates per base pair per generation.</p><p><strong>Optional Arguments</strong></p><ul><li><code>fop::FitOptions = FitOptions(sum(segments), mu, rho)</code>: the fit options, see <a href="#DemoInfer.FitOptions-Tuple{Any, Any, Any}"><code>FitOptions</code></a>.</li><li><code>lo::Int=1</code>: The lowest segment length to be considered in the histogram</li><li><code>hi::Int=50_000_000</code>: The highest segment length to be considered in the histogram</li><li><code>nbins::Int=400</code>: The number of bins to use in the histogram</li><li><code>iters::Int=10</code>: The number of iterations to perform. It might converge earlier</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ArndtLab/DemoInfer.jl/blob/85763d2f4718581c9e647262223d36ddd81445e6/src/corrections.jl#L1-L18">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="DemoInfer.demoinfer-Union{Tuple{E}, Tuple{T}, Tuple{StatsBase.Histogram{T, 1, E}, Int64, FitOptions}} where {T&lt;:Integer, E&lt;:Tuple{AbstractVector{&lt;:Integer}}}" href="#DemoInfer.demoinfer-Union{Tuple{E}, Tuple{T}, Tuple{StatsBase.Histogram{T, 1, E}, Int64, FitOptions}} where {T&lt;:Integer, E&lt;:Tuple{AbstractVector{&lt;:Integer}}}"><code>DemoInfer.demoinfer</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">demoinfer_(h, epochrange, mu, rho, Ltot; kwargs...)</code></pre><p>Infer demographic histories with piece-wise constant epochs where the number of epochs is in <code>epochrange</code>. Takes a histogram as input and the total length of the IBS segments.</p><p>It is much lighter to distribute the histogram than the vector of segments which may also be streamed directly from disk into the histogram.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ArndtLab/DemoInfer.jl/blob/85763d2f4718581c9e647262223d36ddd81445e6/src/corrections.jl#L32-L41">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="DemoInfer.durations-Tuple{FitResult}" href="#DemoInfer.durations-Tuple{FitResult}"><code>DemoInfer.durations</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">durations(fit::FitResult)</code></pre><p>Return the fitted durations of the epochs.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ArndtLab/DemoInfer.jl/blob/85763d2f4718581c9e647262223d36ddd81445e6/src/utils.jl#L67-L71">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="DemoInfer.evd-Tuple{FitResult}" href="#DemoInfer.evd-Tuple{FitResult}"><code>DemoInfer.evd</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">evd(fit::FitResult)</code></pre><p>Return the evidence of the fit.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ArndtLab/DemoInfer.jl/blob/85763d2f4718581c9e647262223d36ddd81445e6/src/utils.jl#L53-L57">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="DemoInfer.get_para-Tuple{FitResult}" href="#DemoInfer.get_para-Tuple{FitResult}"><code>DemoInfer.get_para</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">pars(fit::FitResult)</code></pre><p>Return the parameters of the fit.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ArndtLab/DemoInfer.jl/blob/85763d2f4718581c9e647262223d36ddd81445e6/src/utils.jl#L39-L43">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="DemoInfer.pop_sizes-Tuple{FitResult}" href="#DemoInfer.pop_sizes-Tuple{FitResult}"><code>DemoInfer.pop_sizes</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">pop_sizes(fit::FitResult)</code></pre><p>Return the fitted population sizes, from past to present.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ArndtLab/DemoInfer.jl/blob/85763d2f4718581c9e647262223d36ddd81445e6/src/utils.jl#L60-L64">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="DemoInfer.pre_fit-Union{Tuple{E}, Tuple{T}, Tuple{StatsBase.Histogram{T, 1, E}, Int64, Float64, Float64, Int64, Int64, Number}} where {T&lt;:Integer, E&lt;:Tuple{AbstractVector{&lt;:Integer}}}" href="#DemoInfer.pre_fit-Union{Tuple{E}, Tuple{T}, Tuple{StatsBase.Histogram{T, 1, E}, Int64, Float64, Float64, Int64, Int64, Number}} where {T&lt;:Integer, E&lt;:Tuple{AbstractVector{&lt;:Integer}}}"><code>DemoInfer.pre_fit</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">pre_fit(h::Histogram, nfits, mu, rho, order, Ltot; require_convergence=true)
pre_fit!(fop::FitOptions, h::Histogram, nfits; require_convergence=true)</code></pre><p>Preliminarily fit <code>h</code> with an approximate model of piece-wise constant  epochs for each number of epochs from 1 to <code>nfits</code>.</p><p>With the first signature it initializes the fit options to default. See <a href="#DemoInfer.FitOptions-Tuple{Any, Any, Any}"><code>FitOptions</code></a> for how to specify them. Otherwise it modifies <code>fop</code> in place to adapt it to all the requested epochs. The mutation rate <code>mu</code> and recombination rate <code>rho</code> are assumed to be per base pairs per generation and the total length of the genome <code>Ltot</code> is in base pairs. Return a  vector of <code>FitResult</code>, one for each number of epochs, see also <a href="#DemoInfer.FitResult"><code>FitResult</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ArndtLab/DemoInfer.jl/blob/85763d2f4718581c9e647262223d36ddd81445e6/src/sequential_fit.jl#L133-L149">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="DemoInfer.sds-Tuple{FitResult}" href="#DemoInfer.sds-Tuple{FitResult}"><code>DemoInfer.sds</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">sds(fit::FitResult)</code></pre><p>Return the standard deviations of the parameters of the fit.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ArndtLab/DemoInfer.jl/blob/85763d2f4718581c9e647262223d36ddd81445e6/src/utils.jl#L46-L50">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-nextpage" href="tutorial/">Tutorial »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.15.0 on <span class="colophon-date" title="Wednesday 29 October 2025 06:52">Wednesday 29 October 2025</span>. Using Julia version 1.12.1.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
